util/types/datum.go

```diff
@@ -759,13 +759,26 @@ func (d *Datum) convertToFloat(sc *variable.StatementContext, target *FieldType)
 	default:
 		return invalidConv(d, target.Tp)
 	}
+	var err1 error
+	f, err1 = ProduceFloatWithSpecifiedTp(f, target, sc)
+	if err == nil && err1 != nil {
```

Comment:
Some old [test cases](https://github.com/pingcap/tidb/blob/master/util/types/convert_test.go#L512) depend on this implementation,
so I keep it here.